### ISSUE TITLE
zwave.network_ready does not get triggered in 4.0.0.b1

### ISSUE THREAD

**Issue by electrofloat** [home-assistant/appdaemon#807]
Hi!

It seems that if I use a startup condition like this:
```      
plugin_startup_conditions:
        event: {event_type: zwave.network_ready}
```
then appdaemon is stuck at:
```
2019-11-13 21:48:44.321260 INFO HASS: Connected to Home Assistant 0.101.1
2019-11-13 21:48:44.349081 INFO HASS: Evaluating startup conditions
```
forever. Nothing else is written in the logs after that.

In HomeAssistant I can see that Zwave was ready at:
`2019-11-13 21:49:08 INFO (MainThread) [homeassistant.components.zwave] Z-Wave ready after 27 seconds`

Also it seems that if manually listening for start/ready states in an app like this:
```
    self.listen_event(self.zwave_starting, "zwave.network_start")
    self.listen_event(self.zwave_ready, "zwave.network_ready")
```
the callback functions are never called.

**Comment by ReneTode** [comment]
it is very well possible that HA has changed the event names.
the best way to see if those events are actually there or with another name is to use an app like

```
def intialise(...):
  self.listen_event(self.cb)
def cb(...):
  self.log(data)
```
create and save that app, restart HA and see what events are there from zwave.

if only 1 character in the event_name has changed your CBs will never trigger and AD will never start.

**Comment by electrofloat** [comment]
The constant in homeassistant dev branch is still `zwave.network_ready` https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/zwave/const.py#L74
(and on the 0.101.1), and they seem to fire it here: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/zwave/__init__.py#L521

So I don't think they renamed it, but will try the method you mentioned anyways.

**Comment by electrofloat** [comment]
Hmm.. is it possible that the event system from HomeAssistant doesn't work at all in 4.0.0b1 or did I forget to configure something?
I've just installed Home Assistant and Appdaemon in a demo env, and created an example app:
```
import hassapi as hass
class HelloWorld(hass.Hass):
  def initialize(self):
     self.log("Hello from AppDaemon")
     self.log("You are now ready to run Apps!")
     self.listen_event(self.cb)

  def cb(self, event_name, data, kwargs):
    self.log("Event_name: %s, data: %s", event_name, data)
```
and I only see the `plugin_started` `plugin_stopped` events:
```
2019-11-14 10:23:26.217740 INFO hello_world: Hello from AppDaemon
2019-11-14 10:23:26.218477 INFO hello_world: You are now ready to run Apps!
2019-11-14 10:23:26.221479 INFO hello_world: Event_name: plugin_started, data: {'name': 'HASS'}
2019-11-14 10:25:39.610160 WARNING HASS: Disconnected from Home Assistant, retrying in 5 seconds
2019-11-14 10:25:39.612415 INFO hello_world: Event_name: plugin_stopped, data: {'name': 'HASS'}
2019-11-14 10:25:44.612211 WARNING HASS: Disconnected from Home Assistant, retrying in 5 seconds
2019-11-14 10:25:49.614164 WARNING HASS: Disconnected from Home Assistant, retrying in 5 seconds
2019-11-14 10:25:54.616916 WARNING HASS: Disconnected from Home Assistant, retrying in 5 seconds
2019-11-14 10:25:59.622557 INFO HASS: Connected to Home Assistant 0.101.3
2019-11-14 10:25:59.908087 INFO HASS: Evaluating startup conditions
2019-11-14 10:25:59.913635 INFO AppDaemon: Processing restart for HASS
2019-11-14 10:25:59.913802 INFO AppDaemon: Terminating hello_world
2019-11-14 10:25:59.914065 INFO AppDaemon: Initializing app hello_world using class HelloWorld from module hello
2019-11-14 10:25:59.915654 INFO hello_world: Hello from AppDaemon
2019-11-14 10:25:59.916502 INFO hello_world: You are now ready to run Apps!
2019-11-14 10:25:59.919522 INFO hello_world: Event_name: plugin_started, data: {'name': 'HASS'}
```

**Comment by ReneTode** [comment]
if you got a clean HA it start pretty quick.
so its possible all startup events from HA are already done before AD is listening.
and for sure when you have startup conditions, then the plugin waits untill the startup conditions are met, and only starts listening after that.

**Comment by dlashua** [comment]
@electrofloat With your event listener app running, can you try firing an event directly from Home Assistant and seeing if it shows up in your log?

**Comment by electrofloat** [comment]
> @electrofloat With your event listener app running, can you try firing an event directly from Home Assistant and seeing if it shows up in your log?

Yes, it seems to be working. Also - this is still the demo env - it seems that if I manually start zwave network from the Home Assistant GUI the start event shows up (there won't be any ready events, because in this demo env there are no zwave devices).
So this must be some startup sync issue between Home Assistant and Appdaemon?
(will try it with realy Zwave network later today)
```
2019-11-14 13:59:08.496319 INFO AppDaemon: App initialization complete
2019-11-14 13:59:08.501852 INFO hello_world: Event_name: appd_started, data: {}
2019-11-14 14:02:20.814029 INFO hello_world: Event_name: event_test, data: {}
2019-11-14 14:03:08.307034 INFO hello_world: Event_name: call_service, data: {'domain': 'zwave', 'service': 'start_network', 'service_data': {}}
2019-11-14 14:03:08.309141 INFO hello_world: Event_name: zwave.network_start, data: {}
```

**Comment by ReneTode** [comment]
> So this must be some startup sync issue between Home Assistant and Appdaemon?

probably.
you are listening to events to start AD, but if the event is already fired before that, then AD will never start.

**Comment by electrofloat** [comment]
So.. on a real Z-Wave network this is how it goes.
When restarting Home Assistant this is what I get in appdaemon:
```
2019-11-14 16:31:13.060812 INFO HASS: Connected to Home Assistant 0.101.1
2019-11-14 16:31:13.083221 INFO HASS: Evaluating startup conditions
2019-11-14 16:31:13.090518 INFO AppDaemon: Processing restart for HASS
2019-11-14 16:31:31.643989 INFO persistent_notifications: Event name: zwave.network_complete, data: {}
```
and this is Home Assistant's side:
```
2019-11-14 16:31:31 INFO (Dummy-3) [homeassistant.components.zwave] Z-Wave network is complete. All nodes on the network have been queried
2019-11-14 16:31:31 INFO (MainThread) [homeassistant.components.zwave] Z-Wave ready after 22 seconds
2019-11-14 16:31:31 INFO (SyncWorker_7) [homeassistant.components.zwave] Z-Wave polling interval set to 30000 ms
```

If I manually stop Z-Wave network from Home Assistant frontend->Configuration->Z-Wave->STOP NETWORK and then also manually start this is what I get;
appdaemon:
```
2019-11-14 16:34:10.993825 INFO persistent_notifications: Event name: zwave.network_stop, data: {}
.
.
.
2019-11-14 16:35:17.460678 INFO persistent_notifications: Event name: call_service, data: {'domain': 'zwave', 'service': 'start_network', 'service_data': {}}
2019-11-14 16:35:17.480742 INFO persistent_notifications: Event name: zwave.network_start, data: {}
2019-11-14 16:35:30.500685 INFO persistent_notifications: Event name: zwave.network_complete, data: {}
```
Home Assistant:
```
2019-11-14 16:34:05 INFO (SyncWorker_11) [homeassistant.components.zwave] Stopping Z-Wave network
2019-11-14 16:34:05 INFO (SyncWorker_11) [openzwave] Stop Openzwave network.
2019-11-14 16:34:05 INFO (SyncWorker_11) [openzwave] ZWave configuration written to user directory.
.
.
.
2019-11-14 16:35:30 INFO (Dummy-4) [homeassistant.components.zwave] Z-Wave network is complete. All nodes on the network have been queried
2019-11-14 16:35:31 INFO (MainThread) [homeassistant.components.zwave] Z-Wave ready after 14 seconds
2019-11-14 16:35:31 INFO (SyncWorker_0) [homeassistant.components.zwave] Z-Wave polling interval set to 30000 ms
```

So what I see is that the `zwave.network_ready` event never fires (or at least never reaches appdaemon) even when manually stopping/starting the network from the frontend.

`zwave.network_complete` event always reaches appdaemon (or at least that dummy callback function)

`zwave.network_start` only reaches appdeamon if I manually stop/start Zwave from the frontend.

The issue is that the docs explicitly talks about `hass-plugin-startup-conditions` and explicitly mentions this example:
```
plugin_startup_conditions:
    event: {event_type: zwave.network_ready}
```

which does not work. Also - AFAIK - 4.0 should solve all the startup sync issues with Home Assistant for the exact case I'm having here.

I only wanted to use the startup condition because I run into an issue where I'm setting a Z-Wave config parameter from the `initialize` function but got a log saying:
```
2019-11-14 16:44:05.575337 WARNING HASS: Error calling Home Assistant service default/zwave/set_config_parameter
2019-11-14 16:44:05.575701 WARNING HASS: Code: 400, error: 400: Bad Request
```

**Comment by ReneTode** [comment]
so use zwave.network_complete instead of zwave.network_ready and we will adapt the docs to the change that HA has made.

glad that you found out what they have changed.

**Comment by electrofloat** [comment]
So debugging this further I came to a conclusion that `zwave.network_ready` is not fired on the Home Assistant side.
There's an example automation on `https://www.home-assistant.io/integrations/persistent_notification/`
I used that and saw the same thing as it was with Appdaemon. `zwave.network_ready` was never fired, and start was only triggered when I manually stopped/started Z-Wave. I think I'll create a new issue on Home Assistan's side then. Sorry for the noise!

But the startup conditions got me thinking. Shouldn't be a `timeout` value for both `state` and the `event` startup conditions?
If it is waiting for a state or event it should just move on if it does not happen in timeout. This way we can avoid it being stuck forever. (this is for for example if the machine restarts for whatever reason while noone is at home automations won't ever happen because appdaemon stuck at waiting for the startup condition to be satisfied)

@ReneTode Is it also possible to wait on a certain zwave entity to exist (not just check)? I don't want my initialize function to continue until that entity's parameters can be set.

**Comment by ReneTode** [comment]
i wouldnt hold up my complete AD, and i certainly wouldnt rely on an event from HA.

for things like that i would take the check out of the initialise.
somethink like:
```
def init():
  # set some values
  self.run_in(self.startup,0)

def startup(self):
  if self.entity_exists(zwave_entity):
    # do the initialise stuff
  else:
    self.run_in(self.startup,1)
```

that way all of AD will run as soon as possible, and the app that needs zwave will wait untill the entity exists.

**Comment by electrofloat** [comment]
Unfortunately that does not help, since it seems the zwave entity is already available, but can only be called a set_config_parameter on, after zwave is done initializing itself.
So I resorted to something like this instead:
```
    self.zwave_complete_handle = self.listen_event(self.zwave_complete, "zwave.network_complete")
    self.delayed_startup_handle = self.run_in(self.delayed_startup, 60)

  def zwave_complete(self, event, data, kwargs):
      self.cancel_timer(self.delayed_startup_handle)
      self.log("Z-Wave complete, starting up in 5 seconds")
      self.run_in(self.startup, 5)

  def delayed_startup(self, kwargs):
      self.cancel_listen_event(self.zwave_complete_handle)
      self.log("Z-Wave complete did not fire in 60 seconds, starting up anyway")
      self.startup()
```

This is still not the best solution because now if I restart appdaemon, this app always have to wait 60 seconds to start. And also when zwave.network_complete's callback is called I still have to wait a couple of more seconds, or else it just won't work. God knows why...

Maybe if I can detect why the app initialized itself(Home Assistant restart, or Appdaemon start) I can eliminate the 60 seconds wait on Appdaemon start.

So all in all, not great, not terrible.

**Comment by ReneTode** [comment]
appdeamon has its own startup events, so you can know if AD is restarting.

i am sorry that zwave is such a pain in the butt, all we can do is help you make it as good as possible.

**Comment by electrofloat** [comment]
I forgot about AppDaemon startup events. I'll look into that, thank you!

Also as far as I know they are working on a new method to interact with Z-Wave: https://github.com/cgarwood/python-openzwave-mqtt
That hopefully would eliminate all the pain of Z-Wave startup in the future. We'll see what happens.

I'm closing this issue now since the original issue turned out to be on Home Assistant side.

Thank you!
